quiche/src/recovery/gcongestion/recovery.rs

```diff
@@ -0,0 +1,1002 @@
+            }
+        }
+
+        (time, epoch)
+    }
+
+    fn pto_time_and_space(
+        &self, handshake_status: HandshakeStatus, now: Instant,
+    ) -> (Option<Instant>, packet::Epoch) {
+        let mut duration = self.pto() * (1 << self.pto_count);
+
+        // Arm PTO from now when there are no inflight packets.
+        if self.bytes_in_flight == 0 {
+            if handshake_status.has_handshake_keys {
+                return (Some(now + duration), packet::Epoch::Handshake);
+            } else {
+                return (Some(now + duration), packet::Epoch::Initial);
+            }
+        }
+
+        let mut pto_timeout = None;
+        let mut pto_space = packet::Epoch::Initial;
+
+        // Iterate over all packet number spaces.
+        for &e in packet::Epoch::epochs(
+            packet::Epoch::Initial..=packet::Epoch::Application,
+        ) {
+            if self.epochs[e].pkts_in_flight == 0 {
+                continue;
+            }
+
+            if e == packet::Epoch::Application {
+                // Skip Application Data until handshake completes.
+                if !handshake_status.completed {
+                    return (pto_timeout, pto_space);
+                }
+
+                // Include max_ack_delay and backoff for Application Data.
+                duration +=
+                    self.rtt_stats.max_ack_delay * 2_u32.pow(self.pto_count);
+            }
+
+            let new_time = self.epochs[e]
+                .time_of_last_ack_eliciting_packet
+                .map(|t| t + duration);
+
+            if pto_timeout.is_none() || new_time < pto_timeout {
+                pto_timeout = new_time;
+                pto_space = e;
+            }
+        }
+
+        (pto_timeout, pto_space)
+    }
+
+    fn set_loss_detection_timer(
+        &mut self, handshake_status: HandshakeStatus, now: Instant,
+    ) {
+        if let (Some(earliest_loss_time), _) = self.loss_time_and_space() {
+            // Time threshold loss detection.
+            self.loss_timer.update(earliest_loss_time);
+            return;
+        }
+
+        if self.bytes_in_flight == 0 && handshake_status.peer_verified_address {
+            self.loss_timer.clear();
+            return;
+        }
+
+        // PTO timer.
+        if let (Some(timeout), _) = self.pto_time_and_space(handshake_status, now)
+        {
+            self.loss_timer.update(timeout);
+        }
+    }
+}
+
+impl RecoveryApi for GRecovery {
+    fn lost_count(&self) -> usize {
+        self.lost_count
+    }
+
+    fn bytes_lost(&self) -> u64 {
+        self.bytes_lost
+    }
+
+    fn should_elicit_ack(&self, epoch: packet::Epoch) -> bool {
+        self.epochs[epoch].loss_probes > 0 ||
+            self.outstanding_non_ack_eliciting >=
+                MAX_OUTSTANDING_NON_ACK_ELICITING
+    }
+
+    fn get_acked_frames(&mut self, epoch: packet::Epoch) -> Vec<frame::Frame> {
+        std::mem::take(&mut self.epochs[epoch].acked_frames)
```

Comment:
post: wondering if we can eliminate the use of mem::take since that allocate a new default Vec